### PR TITLE
Hardware unit test

### PR DESCRIPTION
**Pull request by theTechnoWeenie**
@icbat Here's that metaclassing I told you about. :) Lemme know what you think.

### CODE REVIEW

File: hardware/hardware.py [hunk 1]
```diff
@@ -26,8 +26,8 @@ def __init__(self):
 
     def toggle_door(self, door):
         gpio.output(doors[door], 1)
-        sleep(1)
-        gpio.output(doors[door], 0)    
+        time.sleep(1)
```
Comment (icbat) [inline hardware/hardware.py:29]: Out of curiosity, why this change?

Comment (theTechnoWeenie) [inline hardware/hardware.py:29]: This was changed specifically for the unit test.  Basically to use the mock library to mock out the sleep call, I need to be able to pass a mock time object otherwise the unit test is super slow.

File: hardware/test_hardware.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import unittest
```
Comment (icbat) [inline hardware/test_hardware.py:1]: I feel like we need to add some "how to run this" in the readme. Maybe you could start w/ how to use unittest to run the tests?

Comment (theTechnoWeenie) [inline hardware/test_hardware.py:1]: Good call. It's super easy, but I'll at that to the README, basically from any level in the project run `python -m unittest discover` and anything that matches the pattern test*.py will be attempted to be run

Comment (icbat) [inline hardware/test_hardware.py:1]: sweet. and they need the **main** bit to actually run, right?

File: hardware/mockgpio.py [hunk 3]
```diff
@@ -1,21 +1,62 @@
...
-        print(warning_string)
-
-def setup(pin, mode):
-    print_warnings("Setting pin %d to mode %d"%(pin, mode))
-
-def output(pin, value):
-    print_warnings("Setting pin %d to value %s"%(pin,value))
+black_list = ["increment_references", "print_warnings"]
+
+class MockGPIOMetaClass(type):
```
Comment (icbat) [inline hardware/mockgpio.py:3]: This is pretty much black magic to me. Care to give a high-level rundown of how this works?

File: hardware/test_hardware.py (hunk 2 above)
Comment (theTechnoWeenie) [inline hardware/test_hardware.py:1]: Not sure about that part, I'll test with out it and see what happens. I like it there regardless because it allows you to run a single test easily if you are iterating and have a bunch of tests you don't wanna run.

Comment (icbat) [inline hardware/test_hardware.py:1]: That makes sense I suppose. I think the scenario where you specifically
don't want to run tests should be really limited, but I could definitely
see it happening.
- Evan Cobb

In all you do, Good Luck, Have Fun

File: hardware/mockgpio.py (hunk 3 above)
Comment (theTechnoWeenie) [inline hardware/mockgpio.py:3]: Basically, the wizard waves his wand, and the code dances.

Seriously though, the `__metaclass__` is basically a thing that all objects have on them that serves as a registry for where attributes/methods are, and what should be invoked. So by overriding/adding to the meta class you can add hooks in to do tasks on method/attribute access. 

A common use case of this is to add logging to web server. You can add a log message to every time an endpoint is called without adding a bunch of duplicated code to ever endpoint call.

Another common use case is to handle instances where methods don't exist. In python `__getattr__` is called every time a matching attribute cannot be found, and `__getattribute__` (used here) is called every time an attribute is accessed. There is no distinction between methods, and attributes in python except for the presence of `__call__` and for `@classmethod`s the `__func__` attribute.

So basically what is going on here is, I'm adding a hook into the meta class registry to increment a reference counter every time a method is called on `MockGPIO` and also capture the args so that I can assert stuff in the unit tests.  

Did you have any questions that I didn't answer sufficiently?

File: hardware/test_hardware.py (hunk 2 above)
Comment (theTechnoWeenie) [inline hardware/test_hardware.py:1]: Once there is a sufficient number of unit tests, they start to actually take time, and that's when you'll want these. (speaking from having a 2.5 minute build of just unit tests)

File: hardware/mockgpio.py (hunk 3 above)
Comment (icbat) [inline hardware/mockgpio.py:3]: Nope, that's basically it. I got from your comments and the structure how it was going to work, but I was unfamiliar witth the `__metaclass__` so I didn't know how all that magically tied together.
